### PR TITLE
clearing_house: safer liquidations

### PR DESCRIPTION
**Pull request by crispheaney**

### CODE REVIEW

File: programs/clearing_house/src/controller/orders.rs [hunk 1]
```diff
@@ -434,19 +436,15 @@ pub fn fill_order(
     }
 
     // Order fails if it's risk increasing and it brings the user below the initial margin ratio level
```
Comment (0xbigz) [inline programs/clearing_house/src/controller/orders.rs:438]: update comment

Comment (0xbigz) [reply programs/clearing_house/src/controller/orders.rs:438]: Order fails if it's risk increasing and it brings the user collateral below the initial margin requirement

File: programs/clearing_house/src/lib.rs [hunk 2]
```diff
@@ -731,6 +738,7 @@ pub mod clearing_house {
                 funding_rate_history,
                 &ctx.accounts.state.oracle_guard_rails,
                 ctx.accounts.state.funding_paused,
+                Some(mark_price_before),
```
Comment (0xbigz) [inline programs/clearing_house/src/lib.rs:741]: really got this everywhere it could be used huh?

File: programs/clearing_house/src/lib.rs [hunk 3]
```diff
@@ -844,23 +859,22 @@ pub mod clearing_house {
         let mark_price_after = market.amm.mark_price()?;
         let price_oracle = &ctx.accounts.oracle;
 
-        let (oracle_price_after, _oracle_mark_spread_after, oracle_mark_spread_pct_after) =
-            amm::calculate_oracle_mark_spread_pct(
-                &market.amm,
-                &ctx.accounts.oracle,
-                0,
-                clock_slot,
```
Comment (0xbigz) [inline programs/clearing_house/src/lib.rs:852]: and removed clock_slot param passings using oracle data :D

File: programs/clearing_house/src/lib.rs [hunk 4]
```diff
@@ -1116,58 +1131,159 @@ pub mod clearing_house {
...
-        let (total_collateral, unrealized_pnl, base_asset_value, margin_ratio) =
-            calculate_margin_ratio(user, user_positions, &ctx.accounts.markets.load()?)?;
-        if margin_ratio > ctx.accounts.state.margin_ratio_partial {
+        if liquidation_type == LiquidationType::NONE {
             msg!("total_collateral {}", total_collateral);
-            msg!("unrealized_pnl {}", unrealized_pnl);
-            msg!("base_asset_value {}", base_asset_value);
-            msg!("margin ratio {}", margin_ratio);
+            msg!("adjusted_total_collateral {}", adjusted_total_collateral);
+            msg!("margin_requirement {}", margin_requirement);
```
Comment (0xbigz) [inline programs/clearing_house/src/lib.rs:1158]: MEV parsers

File: programs/clearing_house/src/lib.rs [hunk 5]
```diff
@@ -1116,58 +1131,159 @@ pub mod clearing_house {
...
+                    oracle_mark_divergence_after_close,
+                    &state.oracle_guard_rails.price_divergence,
                 )?;
-                if liquidations_blocked {
-                    return Err(ErrorCode::LiquidationsBlockedByOracle.into());
+
+                // if closing pushes outside the oracle mark threshold, don't liquidate
+                if oracle_is_valid && oracle_mark_too_divergent_after_close {
+                    msg!(
+                        "oracle_mark_divergence_after_close {}",
```
Comment (0xbigz) [inline programs/clearing_house/src/lib.rs:1245]: probably would want to indicate the market index here too

File: programs/clearing_house/src/lib.rs [hunk 6]
```diff
@@ -1116,58 +1131,159 @@ pub mod clearing_house {
...
+
+                let oracle_mark_divergence_after_close = if !close_slippage_pct_too_large {
+                    oracle_status
+                        .oracle_mark_spread_pct
+                        .checked_add(close_position_slippage_pct)
+                        .ok_or_else(math_error!())?
+                } else if close_position_slippage_pct > 0 {
+                    oracle_status
+                        .oracle_mark_spread_pct
+                        .checked_add(MAX_LIQUIDATION_SLIPPAGE * 2)
```
Comment (0xbigz) [inline programs/clearing_house/src/lib.rs:1228]: maybe add comment:

approx of constant product curve price impact

File: programs/clearing_house/src/lib.rs [hunk 7]
```diff
@@ -1116,58 +1131,159 @@ pub mod clearing_house {
...
+                // if the oracle is invalid and the mark moves too far from twap, dont liquidate
+                let oracle_is_valid = oracle_status.is_valid;
+                if !oracle_is_valid {
+                    let mark_twap_divergence =
+                        calculate_mark_twap_spread_pct(&market.amm, mark_price_before)?;
+                    let mark_twap_too_divergent =
+                        mark_twap_divergence.unsigned_abs() >= MAX_MARK_TWAP_DIVERGENCE;
+
+                    if mark_twap_too_divergent {
+                        msg!("mark_twap_divergence {}", mark_twap_divergence);
```
Comment (0xbigz) [inline programs/clearing_house/src/lib.rs:1193]: indicate market index here too?

File: programs/clearing_house/src/controller/orders.rs (hunk 1 above)
Comment (crispheaney) [reply programs/clearing_house/src/controller/orders.rs:438]: on it

File: programs/clearing_house/src/lib.rs [hunk 8]
```diff
@@ -1116,58 +1131,159 @@ pub mod clearing_house {
             now,
         )?;
 
+        let LiquidationStatus {
```
Comment (0xbigz) [inline programs/clearing_house/src/lib.rs:1134]: is it worth just putting these in the same order as the variables are listed in the type?

**Pull request merged**